### PR TITLE
[GEOT-7071] GetFeatureInfo in WMSLayer has problems in tranforamation of I,J params

### PR DESCRIPTION
**Pull request by sergitopo**
[GEOS-9770](https://osgeo-org.atlassian.net/browse/GEOS-9770) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

This changes were made to fix GEOS-9770.

- The solution to fix the issue was to use the same CRS source to make the two way tranformations of i,j params.
- There is a test included to validate that real case with the same data that was provided to reproduce the bug.
- Some changes of methods visibility were applied in order to mock/stub behavior of WMSCoverageReader class
- Some other changes are lints and code smells suggested by the IDE in the commit classes to improve the code.
- This merge should be done together the one to geoserver refering to same issue because a test of geoserver will faild and has been adapted for this new version (it is coupled to the behavior of this geotools functionallity) 

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [ ] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java [hunk 1]
```diff
@@ -325,6 +325,14 @@ public GridCoverage2D read(GeneralParameterValue[] parameters)
         return grid;
     }
 
+    public CoordinateReferenceSystem getRequestdEnvelopeCRS() {
```
Comment (aaime) [inline modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java:328]:
```suggestion
    public CoordinateReferenceSystem getRequestedEnvelopeCRS() {
```

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java [hunk 2]
```diff
@@ -237,7 +237,7 @@ private String getCRSWithBoundingBoxes(Layer layer) {
         Map<String, CRSEnvelope> boundingBoxes = layer.getBoundingBoxes();
         return layerSrsSet
                 .stream()
-                .filter(crs -> boundingBoxes.containsKey(crs))
+                .filter(boundingBoxes::containsKey)
```
Comment (aaime) [inline modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java:240]: Please add these in separate commit

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java [hunk 3]
```diff
@@ -349,15 +357,15 @@ GridCoverage2D getMap(
...
+            throw (IOException) new IOException("GetMap failed", e);
         }
     }
 
     /**
      * Sets up a max request with the provided parameters, making sure it is compatible with the
      * layers own native SRS list
      */
-    ReferencedEnvelope initMapRequest(
+    public ReferencedEnvelope initMapRequest(
```
Comment (aaime) [inline modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSCoverageReader.java:368]: Nope, public API is a promise, should not be made public just for testing.

**Review by aaime — CHANGES_REQUESTED** [review]

File: modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSLayer.java [hunk 4]
```diff
@@ -86,10 +86,6 @@ public WMSCoverageReader getReader() {
         return (WMSCoverageReader) this.reader;
     }
 
-    public void setReader(WMSCoverageReader reader) {
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSLayer.java]: Actually no, Sergi just removed a new public method he added in the first iteration of this PR

[the remark this answers is not in the document]
Comment (sergitopo) [reply modules/extension/wms/src/main/java/org/geotools/ows/wms/map/WMSLayer.java]: Yes, the first version of this bug fix code needed to add some public modifiers for testing it, but then I used another test implementation that allowed to leave the public API signature equals to the original.

**Review by aaime — APPROVED** [review]
Looks good, thanks. CLA is covered by GeoSolutions.

**Comment by aaime** [comment]
QA Build has a legit failure:

``
2022-02-15T12:05:16.5127554Z [WARNING] /home/runner/work/geotools/geotools/modules/extension/wms/src/test/java/org/geotools/ows/wms/test/WMSGetFeatureInfoTest.java:[58,73] Double(double) in java.lang.Double has been deprecated
2022-02-15T12:05:16.5128408Z [WARNING] /home/runner/work/geotools/geotools/modules/extension/wms/src/test/java/org/geotools/ows/wms/test/WMSGetFeatureInfoTest.java:[59,80] Double(double) in java.lang.Double has been deprecated
``

Please use ``Double.valueOf(double)`` instead.

**Pull request merged**